### ISSUE TITLE
FormSheet with sheetAllowedDetents: 'fitToContents' doesn't update height dynamically on Android

### ISSUE THREAD

**Issue by SickanK** [software-mansion/react-native-screens#2560]
### Description

The new native stack presentation formSheet with sheetAllowedDetents: 'fitToContents' on Android initially works correctly by adjusting to the content height. However, when the content height changes dynamically, the sheet does not update its height to match. 

I had a short interaction over on [X](https://x.com/kafara_kacper/status/1863560771252195700) and was directed to post an issue here. Love the new formSheets, looking forward to being able to use them with dynamic height adjustment!

### Steps to reproduce

1. Create a new screen with a formSheet presentation mode and sheetAllowedDetents set to fitToContents
2. Add content with a fixed height to the formSheet
3. Open the formSheet from another screen
4. Verify that the initial sheet height matches the fixed content height correctly
5. Change the content's fixed height programmatically
6. Observe that the sheet height remains at its initial size and does not adjust to match the new content height

### Snack or a link to a repository

https://snack.expo.dev/SnPwLOzeISrX17ay1RuQY

### Screens version

4.1.0

### React Native version

0.76.3

### Platforms

Android

### JavaScript runtime

JSC

### Workflow

Expo bare workflow

### Architecture

Fabric (New Architecture)

### Build type

Debug mode

### Device

Real device

### Device model

Nothing Phone (2) (Android 14)

### Acknowledgements

Yes

**Comment by ljukas** [comment]
I heavily use https://github.com/lodev09/react-native-true-sheet for my sheets and they have the same issue regarding auto-sized sheets on Android. How to work around it there is that you also give the largest possible detent as an available size. So you would give ['fitToContents', '1.0'] or similar. This might work here to.

**Comment by kkafar** [comment]
My first impression is that it should be straightforward to do. Gimme few minutes

**Comment by kkafar** [comment]
I can make it jump to the right height, but I can't animate it properly (I don't want to animate it manually, would prefer to delegate it to the bottom sheet behaviour). Will look into it some more

**Comment by kkafar** [comment]
Here's PoC of the changes: https://github.com/software-mansion/react-native-screens/pull/2570

**Comment by SickanK** [comment]
Thanks for working on this! 

Wouldn't the expected behavior mirror what happens when you initially open a sheet without content? What cases would the content disappear where you wouldn't want to instead e.g. close the sheet?

**Comment by RohovDmytro** [comment]
I have encountered a similar issue, but using the iOS Simulator (iPhone, OS v18.2): 

https://github.com/software-mansion/react-native-screens/issues/2688

**Comment by mnelabs-kevin** [comment]
Same here. Commenting to follow along.

**Comment by thebiltheory** [comment]
@kkafar I made a patch with your suggested changes in #2570 and they work quite well. Any chance we can get this PR merged?

[react-native-screens@4.4.0.patch](https://github.com/user-attachments/files/19225220/react-native-screens%404.4.0.patch)

**Comment by sergeymild** [comment]
@kkafar it doesn't work i scrollView changes it is own size 
```
export const ScreenName: React.FC = () => {
  const [count, setCount] = useState(3)
  return (
    <ScrollView>
      {new Array(count).fill(null).map((_, index) => (
        <View
          key={index}
          style={{height: 40, backgroundColor: 'red', borderBottomWidth: 1}}
          nativeID={'someNativeId'}>
          <TouchableOpacity onPress={() => setCount(20)}>
            <Text>{index} - Some text</Text>
          </TouchableOpacity>
        </View>
      ))}
    </ScrollView>
  )
}
```
In this case, if you press on any item and increment the count, the ScrollView's contentSize will be triggered, but the sheet will not update its size.

**Comment by RohovDmytro** [comment]
> [@kkafar](https://github.com/kkafar) it doesn't work i scrollView changes it is own size
> 
> ```
> export const ScreenName: React.FC = () => {
>   const [count, setCount] = useState(3)
>   return (
>     <ScrollView>
>       {new Array(count).fill(null).map((_, index) => (
>         <View
>           key={index}
>           style={{height: 40, backgroundColor: 'red', borderBottomWidth: 1}}
>           nativeID={'someNativeId'}>
>           <TouchableOpacity onPress={() => setCount(20)}>
>             <Text>{index} - Some text</Text>
>           </TouchableOpacity>
>         </View>
>       ))}
>     </ScrollView>
>   )
> }
> ```
> 
> In this case, if you press on any item and increment the count, the ScrollView's contentSize will be triggered, but the sheet will not update its size.

I confirm this being the case in the latest release (4.10.0)

**Comment by DvzZDev** [comment]
@kkafar Any progress on this issue? It’s been a while, and this is a feature that would be greatly appreciated on Android. Thanks guys for all the great work you do.

**Comment by t0maboro** [comment]
We're working on it; you can follow the progress here: https://github.com/software-mansion/react-native-screens/pull/3484 . It's currently under review.

**Comment by marcel-happyfloat** [comment]
@t0maboro I had this issue before, then your change fixed it, but now it seems to be back after upgrading from "4.20.0", to "4.23.0", (and RN 81 to 83, Expo SDK 54 -> 55 beta, ...). it seems to be broken again on android. 
It gets stuck for dynamic content, still works for static content.

**Comment by NicoKevin** [comment]
> [@t0maboro](https://github.com/t0maboro) I had this issue before, then your change fixed it, but now it seems to be back after upgrading from "4.20.0", to "4.23.0", (and RN 81 to 83, Expo SDK 54 -> 55 beta, ...). it seems to be broken again on android. It gets stuck for dynamic content, still works for static content.

Reproduce this issue on expo 55 and react native screens 4.23.0. The dynamic content doesn't work always.

**Comment by t0maboro** [comment]
Hi,
I checked out internal examples, and I haven't noticed any issues. Could you provide more details, e.g., share the FormSheet content or provide a repro? Is there any difference if you're using `sheetResizeAnimationEnabled: false` or not?

**Comment by marcel-happyfloat** [comment]
@t0maboro Yep — setting sheetResizeAnimationEnabled: false fixes it, thanks!

It still feels like the sheet “animates in” on open (which is nice), but it no longer animates between size changes (which is expected with this flag) after it’s already visible  — works for my use case.

In my case I always fetch data during mount, so the content (and height) changes shortly after opening. Without this flag, the sheet sometimes stays stuck at the initial height (before the fetched data renders). And when I close it, it briefly flickers to the correct height right before dismissing.

**Comment by iankberry** [comment]
I was able to create a reproduction of this issue if that's helpful. I can't speak to the exact conditions of @marcel-happyfloat's app, but in my case it centers around having sheet content that grows in an animated container. So it seems that the animation within the sheet might be interfering with the sheet animation. In this app, if you disable the sheet animation with `sheetResizeAnimationEnabled` OR field the group container with the `animateHeight` prop, the issue no longer occurs.

To reproduce:
1. Check out https://github.com/iankberry/android-formsheet-fittocontents-issue
2. Run `yarn install && yarn android`
3. Click the Open Formsheet (Growing Content) button

https://github.com/user-attachments/assets/9869706c-4c10-401a-a435-cfce5b493506

**Comment by t0maboro** [comment]
@marcel-happyfloat  @iankberry 
Thanks for sharing details.

I need to start with some technical context on why this happens. The issue is that this setup triggers two animations to run simultaneously:
- default native animation from react-native-screens, which operates directly on native translations (performing direct native view updates under the hood).
- your custom animation in RN app, that's handled inside the RN core, committing layout changes to Yoga.

This creates two separate sources of truth, which are trying to update the layout at the same time. It is practically impossible to perfectly synchronize direct native frame updates with Yoga layout commits from RN animations.

The situation seems to be similar when you're dynamically changing the size on the sheet entering - RN sends a layout update, and our default animation is working in the same time.

---

Given that, this is the exact scenario for which we introduced the `sheetResizeAnimationEnabled` prop. It serves as an opt-out from the default animation we provide. If you intend to use your own custom animations to resize the content (Animated API, react-native-reanimated etc.) or if some layout updates may interfere with the native animation state, it's recommended to use this prop to disable the default native animation and prevent latout issues.

**Comment by iankberry** [comment]
@t0maboro Makes sense, thanks for the detailed explanation. I can certainly see why this would cause issues and in my case I just disabled the custom animation because it wasn't really contributing much of a benefit.

**Comment by marcel-happyfloat** [comment]
@t0maboro I do not have animations on my screens, at least not that I'm aware off, also it works fine on iOS in any case - as it did on android with 4.20.0 (and other package downgrades). I need to finish up my new release first, before I find more time trying to reproduce this with a minimal repo... sorry!

**Comment by t0maboro** [comment]
@marcel-happyfloat 

that's might be the case I mentioned here

https://github.com/software-mansion/react-native-screens/issues/2560#issuecomment-3991737608
> The situation seems to be similar when you're dynamically changing the size on the sheet entering - RN sends a layout update, and our default animation is working in the same time.

if the layout changes before the sheet has entered, there are two different sources that are providing their frames, what may doesn't work well with our sheet resize animation which is based on the difference between old and new height

anyway, if it touches only 4.20+, but on 4.20 it works fine - that's sth I'm not expecting, so any assistance with reproducing that would be really helpful

**Comment by marcel-happyfloat** [comment]
@t0maboro I will try downgrading to this version again, without touching any other packages and then I will report back. Tomorrow or the day after

**Comment by marcel-happyfloat** [comment]
@t0maboro I downgraded only this package to 4.20, and it started working again. But I noticed it doesn’t open as smoothly as  with animations disabled, so it looks like it still gets stuck in the wrong position at first (30ms~), then at least updates visually. So the issue seems to happen earlier (and not between these versions), but this version appears to handle it a bit better...

Edit: Maybe my network was just slower loading my dynamic data today... so maybe the issue is between these versions... no idea.

**Comment by jonstuebe** [comment]
My findings is that the `SafeAreaView` from `react-native-safe-area-context`was causing all my problems. Instead of using that in tandem with a `ScrollView`, I'm simply using the `useSafeAreaInsets` and passing `insets.bottom` to my `SafeAreaView` `style` prop and that solved the issue. Buttery smooth animations!
